### ISSUE TITLE
Manage nested Pulumi stack updates from within a Pulumi program

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#2209]
It's a best practice in Pulumi to break larger stacks down into separate stacks which version independently and have different owners, lifecycles, access control, etc.

We [recently added](https://github.com/pulumi/pulumi/pull/2180) the ability to reference a stack using `pulumi.StackReference`, which makes it easy to refer to other stacks which have been deployed independently.

However, there is also often a desire to have a Pulumi program which sits above several others and coordinates deployment of each of the sub-components.  

We have imagined that we could support this via something like the following:

```typescript
import * as pulumi from "@pulumi/pulumi";

let baseNetwork = new pulumi.Stack("network", {
  path: "./network",
});

let app = new pulumi.Stack("app", {
  path: "./app",
  config: {
    networkStackName: baseNetwork.stackName,
  },
});
```

Deploying the Pulumi program above would deploy stacks for the "network" and "app" projects.  Updating the stack above would then lead to updating (as necessary) each of those stacks.  Notably, the reference between the stacks can be passed via `config` such that changes cascade correctly across these sub-stacks.

This could work at two levels.  Most obvious would be just having the Pulumi engine that is driving deployment of the program be able to deploy each of the sub-stacks all as part of the same engine process.  Alternatively, this could rely on some remote-deployment model.

Other notes:
1. Path could either be a path on disk or a git repo reference (like we support in `pulumi up` today)
1. Presumably the args on this `Stack` would include everything that `pulumi up` supports today (plus potentially anything `pulumi stack init` supports)
1. The lifecycle of this "Stack" Resource includes effectively `pulumi stack init && pulumi update` on Create, `pulumi update` on Update `pulumi destroy && pulumi stack rm` on Destroy and `pulumi stack` on Read.
1. We already have a `pulumi.Stack` today which is the (currently implicit) root resource of a Stack.  The resource described here is in many ways the same thing, but with the ability to explicitly create one.  We would need to decide if and how we can unify these concepts.
1. Worth noting that these sub-projects could be defined and instantiated entirely indpendently of this umbrella stack - for example, developers could stand up dev stacks of just `app` and configure them themselves.  That offers flexibility to recompose these separate projects together in multiple different "umbrella" projects, and to have different environments stand up different sub-stacks.  Only the specific stack instantiation of the sub-project used by the umbrella project is explicitly managed by the umbrella project.  This generally feels like it would be quite flexible.

**Comment by gobengo** [comment]
This has been rolled over from each milestone for several months . Is it really ongoing or is it blocked somehow? Will I be able to use a feature like this any time soon? just wondering as I think about how to set up big projects and automate deployment in a gitopsy way.

**Comment by lukehoban** [comment]
@gobengo This idea here is still at a somewhat early stage - we haven't started any work on this yet.  I'd love to hear feedback on concrete scenarios where something like this would be useful, and any feedback on the specific suggestion outlined above.

**Comment by Place1** [comment]
This looks like a really useful feature. My current use-case is to provide a shared resource stack which contains a GKE cluster (and a few other bits and pieces), then a separate stack per environment which consist of different kubernetes resources being deployed to the shared cluster.

Ideally, we could set this up using 1 pulumi program because it's all part of the same logical project and development team.

Perhaps pulumi could support specifying which stack a resource belonged too in the pulumi program itself. e.g.

```typescript
const sharedStack = new pulumi.Stack("shared");

export const cluster = new gcp.container.Cluster('cluster', {
  // ...
}, {
  stack: sharedStack,
});
```

This approach might simplify `StackReference`s for monolithic pulumi programs that deploy multiple stacks. Other parts of the pulumi program could import this `cluster` to use `cluster.endpoint` directly via a ts-import, and pulumi knows which stack to get the output from because it's declared in the program.

**Comment by GoodMirek** [comment]
Is there any plan for this feature?
E.g. it will be implemented this year, next year or no plan to deliver it in the next 18 months.

**Comment by 1oglop1** [comment]
I'd like this to be a feature it seem to be very similar to the approach of AWS CDK.

**Comment by tim-ethitrade** [comment]
@GoodMirek It looks like it's been put on [the backlog](https://github.com/pulumi/pulumi/wiki/Roadmap#backlog) if that helps.

**Comment by jagu-sayan** [comment]
New backlog : https://github.com/orgs/pulumi/projects/44

**Comment by stefanhenseler** [comment]
@lukehoban I was looking into a similar use case. As we now have the pulumi deployments feature, would it make sense to revisit this topic? It would be neat to have a pulumi deployment resource and provider using the deployment API as an interface https://www.pulumi.com/docs/reference/deployments-rest-api/. This would allow users to create a Deployment of a Stack via the Pulumi Service Deployment from within a program. The community could also create other deployment provider implementations. I thought of creating a Deployment provider for deploying a program in a customer tenant via azure lighthouse using a container instance and managed identity to access the customer's data plane. This would abstract the deployment of a stack and make the outputs available within the current program without leaving it. ideally, the deployment resource should behave like any other Resource. Is this something you were looking in to anyways, or does it make sense that I give it a go in a little side project?

**Comment by ringods** [comment]
@stefanhenseler if I read your feedback correctly, I think the `Remote Workspace` support in our Automation API is what you want.

https://www.pulumi.com/docs/intro/pulumi-service/deployments/#automation-via-the-pulumi-automation-api

Have a look at it and let us know if this supports your needs.

**Comment by stefanhenseler** [comment]
Yeah, no not really. The Remote Workspace is to run a program in Pulumi Service Deployment, as far as I understand. But I'm looking for a way to run a program nested in another program on any kind of remote execution environment like for example Azure Container Instance or Google Workflows etc.

**Comment by blampe** [comment]
Something I've done in the past:

In the "root" stack's config, declare something like

```
$ cat root/Pulumi.production.yaml
...
dependsOn:
- pulumi/foo/production
- pulumi/bar/production
```

These dependent stacks can have their own dependencies:
```
$ cat foo/Pulumi.production.yaml
...
dependsOn:
- pulumi/baz/production
```

(Statically declaring `dependsOn` isn't strictly necessary -- you can find these dependencies by inspecting outputs for StackReferences -- but it addresses the problem where a stack is deployed for the first time and doesn't yet have a StackReference in its outputs.)


Then to deploy "root" use a small automation API program that traverses the tree of `dependsOn` dependencies and deploys things in topological order (and in parallel, where possible).

**Comment by EvanBoyle** [comment]
Today we launched support for automatically updating dependent stacks via Pulumi Deployments.

https://www.pulumi.com/blog/dependent-stack-updates/

This is accomplished via two new features within Pulumi Cloud:

1. [Deployment Webhook Destinations](https://www.pulumi.com/docs/pulumi-cloud/webhooks/#deployment-webhooks) for triggering downstream stack updates automatically
2. [The Pulumi Auto Deploy Package](https://www.pulumi.com/registry/packages/auto-deploy) for easily expressing dependencies and configuring downstream updates via Deployment Webhooks.

The Auto Deploy Package makes it easy to express dependencies between stacks and Pulumi Deployments takes care of keeping them up to date:

```ts
export const b = new autodeploy.AutoDeployer("auto-deployer-b", {
    organization,
    project,
    stack: "b",
    downstreamRefs: [d.ref, e.ref, f.ref],

});

export const a = new autodeploy.AutoDeployer("auto-deployer-a", {
    organization,
    project,
    stack: "a",
    downstreamRefs: [b.ref, c.ref],
});
```

Please feel free to submit feedback and open feature requests via https://github.com/pulumi/pulumi-cloud-requests

Happy building! 🚀

**Comment by geowalrus4gh** [comment]
waiting for the same to non cloud, open-source users🙏

**Comment by brianzilliz** [comment]
Is this still planned? Being able to control the provisioning of other pulumi resources within a pulimi program itself would be a great feature.

Something like AWS's [CDK Pipelines](https://docs.aws.amazon.com/cdk/v2/guide/cdk_pipeline.html) allows you to provision your infrastructure as well as the deployment pipeline for that infrastructure using the same IaC library (CDK), it would be great if Pulumi could do this as well. As of now we have to configure the deployment pipeline outside of Pulumi so we lose out on Pulumi's benefits there.

I'm aware of the Deployments feature although I am using open-source Pulumi. Also the Deployments feature seems only to concern itself with dependencies between stacks, nothing with actually provisioning the stacks themselves, which still needs to be done outside Pulumi.
